File: domains/test/python/src/core/validators.py
"""
Validation utilities for the analytics platform.

This module provides validators for data types, schemas, configurations,
and other platform components.
"""

import re
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Optional, Type, Union
from pathlib import Path

from .exceptions import ValidationError, SchemaError


@dataclass
class ValidationRule:
    """Represents a validation rule."""
    
    name: str
    validator: Callable[[Any], bool]
    error_message: str
    required: bool = True
    
    def validate(self, value: Any) -> bool:
        """Apply the validation rule.
        
        Args:
            value: Value to validate
            
        Returns:
            True if validation passes
            
        Raises:
            ValidationError: If validation fails
        """
        if value is None and not self.required:
            return True
        
        if value is None and self.required:
            raise ValidationError(f"{self.name}: Value is required")
        
        try:
            result = self.validator(value)
            if not result:
                raise ValidationError(f"{self.name}: {self.error_message}")
            return True
        except Exception as e:
            if isinstance(e, ValidationError):
                raise
            raise ValidationError(f"{self.name}: Validation failed - {str(e)}")


class BaseValidator(ABC):
    """Abstract base class for validators."""
    
    def __init__(self, strict: bool = False):
        """Initialize validator.
        
        Args:
            strict: If True, raise exceptions on validation failure
        """
        self.strict = strict
        self.errors: List[str] = []
    
    @abstractmethod
    def validate(self, value: Any) -> bool:
        """Validate a value.
        
        Args:
            value: Value to validate
            
        Returns:
            True if validation passes
        """
        pass
    
    def __call__(self, value: Any) -> bool:
        """Make validator callable.
        
        Args:
            value: Value to validate
            
        Returns:
            True if validation passes
        """
        return self.validate(value)
    
    def add_error(self, error: str) -> None:
        """Add a validation error.
        
        Args:
            error: Error message
        """
        self.errors.append(error)
        if self.strict:
            raise ValidationError(error)
    
    def clear_errors(self) -> None:
        """Clear all validation errors."""
        self.errors.clear()
    
    def get_errors(self) -> List[str]:
        """Get list of validation errors.
        
        Returns:
            List of error messages
        """
        return self.errors.copy()


class TypeValidator(BaseValidator):
    """Validator for checking data types."""
    
    def __init__(self, expected_type: Type, strict: bool = False):
        """Initialize type validator.
        
        Args:
            expected_type: Expected data type
            strict: If True, raise exceptions on validation failure
        """
        super().__init__(strict)
        self.expected_type = expected_type
    
    def validate(self, value: Any) -> bool:
        """Validate that value matches expected type.
        
        Args:
            value: Value to validate
            
        Returns:
            True if type matches
        """
        if not isinstance(value, self.expected_type):
            self.add_error(
                f"Expected type {self.expected_type.__name__}, "
                f"got {type(value).__name__}"
            )
            return False
        return True


class RangeValidator(BaseValidator):
    """Validator for numeric ranges."""
    
    def __init__(self, min_value: Optional[float] = None,
                 max_value: Optional[float] = None, strict: bool = False):
        """Initialize range validator.
        
        Args:
            min_value: Minimum allowed value
            max_value: Maximum allowed value
            strict: If True, raise exceptions on validation failure
        """
        super().__init__(strict)
        self.min_value = min_value
        self.max_value = max_value
    
    def validate(self, value: Any) -> bool:
        """Validate that value is within range.
        
        Args:
            value: Value to validate
            
        Returns:
            True if value is within range
        """
        if not isinstance(value, (int, float)):
            self.add_error(f"Value must be numeric, got {type(value).__name__}")
            return False
        
        if self.min_value is not None and value < self.min_value:
            self.add_error(f"Value {value} is below minimum {self.min_value}")
            return False
        
        if self.max_value is not None and value > self.max_value:
            self.add_error(f"Value {value} is above maximum {self.max_value}")
            return False
        
        return True


class PatternValidator(BaseValidator):
    """Validator for string patterns using regex."""
    
    def __init__(self, pattern: str, strict: bool = False):
        """Initialize pattern validator.
        
        Args:
            pattern: Regular expression pattern
            strict: If True, raise exceptions on validation failure
        """
        super().__init__(strict)
        self.pattern = re.compile(pattern)
    
    def validate(self, value: Any) -> bool:
        """Validate that value matches pattern.
        
        Args:
            value: Value to validate
            
        Returns:
            True if value matches pattern
        """
        if not isinstance(value, str):
            self.add_error(f"Value must be string, got {type(value).__name__}")
            return False
        
        if not self.pattern.match(value):
            self.add_error(f"Value '{value}' does not match pattern {self.pattern.pattern}")
            return False
        
        return True


class SchemaValidator(BaseValidator):
    """Validator for dictionary schemas."""
    
    def __init__(self, schema: Dict[str, Union[Type, ValidationRule]], 
                 strict: bool = False):
        """Initialize schema validator.
        
        Args:
            schema: Dictionary defining expected structure
            strict: If True, raise exceptions on validation failure
        """
        super().__init__(strict)
        self.schema = schema
    
    def validate(self, value: Any) -> bool:
        """Validate that value matches schema.
        
        Args:
            value: Value to validate
            
        Returns:
            True if value matches schema
        """
        if not isinstance(value, dict):
            self.add_error(f"Value must be dictionary, got {type(value).__name__}")
            return False
        
        valid = True
        
        # Check required fields
        for field, rule in self.schema.items():
            if isinstance(rule, ValidationRule):
                if rule.required and field not in value:
                    self.add_error(f"Required field '{field}' is missing")
                    valid = False
                elif field in value:
                    try:
                        rule.validate(value[field])
                    except ValidationError as e:
                        self.add_error(f"Field '{field}': {str(e)}")
                        valid = False
            elif isinstance(rule, type):
                if field not in value:
                    self.add_error(f"Required field '{field}' is missing")
                    valid = False
                elif not isinstance(value[field], rule):
                    self.add_error(
                        f"Field '{field}' expected type {rule.__name__}, "
                        f"got {type(value[field]).__name__}"
                    )
                    valid = False
        
        # Check for unexpected fields in strict mode
        if self.strict:
            unexpected = set(value.keys()) - set(self.schema.keys())
            if unexpected:
                self.add_error(f"Unexpected fields: {unexpected}")
                valid = False
        
        return valid


class FileValidator(BaseValidator):
    """Validator for file paths and properties."""
    
    def __init__(self, must_exist: bool = True, 
                 allowed_extensions: Optional[List[str]] = None,
                 max_size_bytes: Optional[int] = None, strict: bool = False):
        """Initialize file validator.
        
        Args:
            must_exist: If True, file must exist
            allowed_extensions: List of allowed file extensions
            max_size_bytes: Maximum file size in bytes
            strict: If True, raise exceptions on validation failure
        """
        super().__init__(strict)
        self.must_exist = must_exist
        self.allowed_extensions = allowed_extensions
        self.max_size_bytes = max_size_bytes
    
    def validate(self, value: Any) -> bool:
        """Validate file path.
        
        Args:
            value: File path to validate
            
        Returns:
            True if file path is valid
        """
        if isinstance(value, str):
            path = Path(value)
        elif isinstance(value, Path):
            path = value
        else:
            self.add_error(f"Value must be string or Path, got {type(value).__name__}")
            return False
        
        if self.must_exist and not path.exists():
            self.add_error(f"File does not exist: {path}")
            return False
        
        if path.exists() and not path.is_file():
            self.add_error(f"Path is not a file: {path}")
            return False
        
        if self.allowed_extensions:
            if path.suffix not in self.allowed_extensions:
                self.add_error(
                    f"File extension '{path.suffix}' not in allowed "
                    f"extensions: {self.allowed_extensions}"
                )
                return False
        
        if self.max_size_bytes and path.exists():
            size = path.stat().st_size
            if size > self.max_size_bytes:
                self.add_error(
                    f"File size {size} bytes exceeds maximum "
                    f"{self.max_size_bytes} bytes"
                )
                return False
        
        return True


class CompositeValidator(BaseValidator):
    """Validator that combines multiple validators."""
    
    def __init__(self, validators: List[BaseValidator], 
                 require_all: bool = True, strict: bool = False):
        """Initialize composite validator.
        
        Args:
            validators: List of validators to apply
            require_all: If True, all validators must pass
            strict: If True, raise exceptions on validation failure
        """
        super().__init__(strict)
        self.validators = validators
        self.require_all = require_all
    
    def validate(self, value: Any) -> bool:
        """Validate using all sub-validators.
        
        Args:
            value: Value to validate
            
        Returns:
            True if validation passes
        """
        results = []
        
        for validator in self.validators:
            try:
                result = validator.validate(value)
                results.append(result)
                if not result:
                    self.errors.extend(validator.get_errors())
                    if self.require_all:
                        return False
            except ValidationError as e:
                self.add_error(str(e))
                if self.require_all:
                    return False
        
        if self.require_all:
            return all(results)
        else:
            return any(results)


# Utility functions for common validations

def validate_email(email: str) -> bool:
    """Validate email address format.
    
    Args:
        email: Email address to validate
        
    Returns:
        True if email format is valid
    """
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def validate_url(url: str) -> bool:
    """Validate URL format.
    
    Args:
        url: URL to validate
        
    Returns:
        True if URL format is valid
    """
    pattern = r'^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&\/\/=]*)$'
    return bool(re.match(pattern, url))


def validate_json_schema(data: Dict, schema: Dict) -> bool:
    """Validate data against JSON schema.
    
    Args:
        data: Data to validate
        schema: JSON schema definition
        
    Returns:
        True if data matches schema
    """
    # Simplified JSON schema validation
    validator = SchemaValidator(schema, strict=True)
    return validator.validate(data)

File: domains/test/javascript/src/core/validators.js
// core/validators.js
/**
 * Validation utilities for the analytics platform.
 * 
 * This module provides validators for data types, schemas, configurations,
 * and other platform components.
 */

import path from 'path';
import fs from 'fs';
import { ValidationError, SchemaError } from './exceptions.js';

/**
 * Represents a validation rule
 */
export class ValidationRule {
    constructor(name, validator, errorMessage, required = true) {
        this.name = name;
        this.validator = validator;
        this.errorMessage = errorMessage;
        this.required = required;
    }

    validate(value) {
        if (value === null || value === undefined) {
            if (this.required) {
                throw new ValidationError(`${this.name}: Value is required`);
            }
            return true;
        }

        try {
            const result = this.validator(value);
            if (!result) {
                throw new ValidationError(`${this.name}: ${this.errorMessage}`);
            }
            return true;
        } catch (error) {
            if (error instanceof ValidationError) {
                throw error;
            }
            throw new ValidationError(`${this.name}: Validation failed - ${error.message}`);
        }
    }
}

/**
 * Abstract base class for validators
 */
export class BaseValidator {
    constructor(strict = false) {
        this.strict = strict;
        this.errors = [];
    }

    /**
     * Validate a value
     * @abstract
     */
    validate(value) {
        throw new Error('Method validate() must be implemented');
    }

    /**
     * Make validator callable as a function
     */
    call(value) {
        return this.validate(value);
    }

    addError(error) {
        this.errors.push(error);
        if (this.strict) {
            throw new ValidationError(error);
        }
    }

    clearErrors() {
        this.errors = [];
    }

    getErrors() {
        return [...this.errors];
    }

    hasErrors() {
        return this.errors.length > 0;
    }
}

/**
 * Validator for checking data types
 */
export class TypeValidator extends BaseValidator {
    constructor(expectedType, strict = false) {
        super(strict);
        this.expectedType = expectedType;
    }

    validate(value) {
        const actualType = typeof value;
        
        // Handle special cases
        if (this.expectedType === 'array' && !Array.isArray(value)) {
            this.addError(`Expected array, got ${actualType}`);
            return false;
        }
        
        if (this.expectedType === 'null' && value !== null) {
            this.addError(`Expected null, got ${actualType}`);
            return false;
        }
        
        if (this.expectedType !== 'array' && this.expectedType !== 'null') {
            if (actualType !== this.expectedType) {
                this.addError(`Expected type ${this.expectedType}, got ${actualType}`);
                return false;
            }
        }
        
        return true;
    }
}

/**
 * Validator for numeric ranges
 */
export class RangeValidator extends BaseValidator {
    constructor(minValue = null, maxValue = null, strict = false) {
        super(strict);
        this.minValue = minValue;
        this.maxValue = maxValue;
    }

    validate(value) {
        if (typeof value !== 'number' && typeof value !== 'bigint') {
            this.addError(`Value must be numeric, got ${typeof value}`);
            return false;
        }

        const numValue = Number(value);

        if (this.minValue !== null && numValue < this.minValue) {
            this.addError(`Value ${numValue} is below minimum ${this.minValue}`);
            return false;
        }

        if (this.maxValue !== null && numValue > this.maxValue) {
            this.addError(`Value ${numValue} is above maximum ${this.maxValue}`);
            return false;
        }

        return true;
    }
}

/**
 * Validator for string patterns using regex
 */
export class PatternValidator extends BaseValidator {
    constructor(pattern, strict = false) {
        super(strict);
        this.pattern = pattern instanceof RegExp ? pattern : new RegExp(pattern);
    }

    validate(value) {
        if (typeof value !== 'string') {
            this.addError(`Value must be string, got ${typeof value}`);
            return false;
        }

        if (!this.pattern.test(value)) {
            this.addError(`Value '${value}' does not match pattern ${this.pattern}`);
            return false;
        }

        return true;
    }
}

/**
 * Validator for object schemas
 */
export class SchemaValidator extends BaseValidator {
    constructor(schema, strict = false) {
        super(strict);
        this.schema = schema;
    }

    validate(value) {
        if (typeof value !== 'object' || value === null) {
            this.addError(`Value must be object, got ${typeof value}`);
            return false;
        }

        let valid = true;

        // Check required fields
        for (const [field, rule] of Object.entries(this.schema)) {
            if (rule instanceof ValidationRule) {
                if (rule.required && !(field in value)) {
                    this.addError(`Required field '${field}' is missing`);
                    valid = false;
                } else if (field in value) {
                    try {
                        rule.validate(value[field]);
                    } catch (error) {
                        this.addError(`Field '${field}': ${error.message}`);
                        valid = false;
                    }
                }
            } else if (typeof rule === 'function') {
                // Simple type check
                if (!(field in value)) {
                    this.addError(`Required field '${field}' is missing`);
                    valid = false;
                } else if (rule.name && typeof value[field] !== rule.name.toLowerCase()) {
                    this.addError(
                        `Field '${field}' expected type ${rule.name}, ` +
                        `got ${typeof value[field]}`
                    );
                    valid = false;
                }
            }
        }

        // Check for unexpected fields in strict mode
        if (this.strict) {
            const schemaKeys = new Set(Object.keys(this.schema));
            const unexpected = Object.keys(value).filter(key => !schemaKeys.has(key));
            if (unexpected.length > 0) {
                this.addError(`Unexpected fields: ${unexpected.join(', ')}`);
                valid = false;
            }
        }

        return valid;
    }
}

/**
 * Validator for file paths and properties
 */
export class FileValidator extends BaseValidator {
    constructor(options = {}, strict = false) {
        super(strict);
        this.mustExist = options.mustExist ?? true;
        this.allowedExtensions = options.allowedExtensions || null;
        this.maxSizeBytes = options.maxSizeBytes || null;
    }

    validate(value) {
        let filePath;
        
        if (typeof value === 'string') {
            filePath = path.resolve(value);
        } else if (value && typeof value === 'object' && value.path) {
            filePath = path.resolve(value.path);
        } else {
            this.addError(`Value must be string or path object, got ${typeof value}`);
            return false;
        }

        if (this.mustExist && !fs.existsSync(filePath)) {
            this.addError(`File does not exist: ${filePath}`);
            return false;
        }

        if (fs.existsSync(filePath)) {
            const stats = fs.statSync(filePath);
            
            if (!stats.isFile()) {
                this.addError(`Path is not a file: ${filePath}`);
                return false;
            }

            if (this.allowedExtensions) {
                const ext = path.extname(filePath);
                if (!this.allowedExtensions.includes(ext)) {
                    this.addError(
                        `File extension '${ext}' not in allowed extensions: ` +
                        `${this.allowedExtensions.join(', ')}`
                    );
                    return false;
                }
            }

            if (this.maxSizeBytes && stats.size > this.maxSizeBytes) {
                this.addError(
                    `File size ${stats.size} bytes exceeds maximum ${this.maxSizeBytes} bytes`
                );
                return false;
            }
        }

        return true;
    }
}

/**
 * Composite validator that combines multiple validators
 */
export class CompositeValidator extends BaseValidator {
    constructor(validators, requireAll = true, strict = false) {
        super(strict);
        this.validators = validators;
        this.requireAll = requireAll;
    }

    validate(value) {
        const results = [];

        for (const validator of this.validators) {
            try {
                const result = validator.validate(value);
                results.push(result);
                
                if (!result) {
                    this.errors.push(...validator.getErrors());
                    if (this.requireAll) {
                        return false;
                    }
                }
            } catch (error) {
                this.addError(error.message);
                if (this.requireAll) {
                    return false;
                }
            }
        }

        if (this.requireAll) {
            return results.every(r => r);
        } else {
            return results.some(r => r);
        }
    }
}

// Utility validation functions

/**
 * Validate email address format
 */
export function validateEmail(email) {
    const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return pattern.test(email);
}

/**
 * Validate URL format
 */
export function validateURL(url) {
    try {
        new URL(url);
        return true;
    } catch {
        return false;
    }
}

/**
 * Validate ISO date string
 */
export function validateISODate(dateStr) {
    const date = new Date(dateStr);
    return date instanceof Date && !isNaN(date) && date.toISOString() === dateStr;
}

/**
 * Validate JSON string
 */
export function validateJSON(jsonStr) {
    try {
        JSON.parse(jsonStr);
        return true;
    } catch {
        return false;
    }
}

/**
 * Create a custom validator function
 */
export function createValidator(validationFn, errorMessage = 'Validation failed') {
    return class CustomValidator extends BaseValidator {
        validate(value) {
            try {
                const result = validationFn(value);
                if (!result) {
                    this.addError(errorMessage);
                    return false;
                }
                return true;
            } catch (error) {
                this.addError(`${errorMessage}: ${error.message}`);
                return false;
            }
        }
    };
}

/**
 * Validate data against JSON schema (simplified)
 */
export function validateJSONSchema(data, schema) {
    const validator = new SchemaValidator(schema, true);
    return validator.validate(data);
}

/**
 * Chain multiple validators
 */
export function chain(...validators) {
    return new CompositeValidator(validators, true);
}

/**
 * At least one validator must pass
 */
export function oneOf(...validators) {
    return new CompositeValidator(validators, false);
}